src/index.js

```diff
@@ -0,0 +1,236 @@
+'use strict';
+
+var redis = require('redis');
+var Scripty = require('node-redis-scripty');
+var EventEmitter = require('events').EventEmitter;
+
+/**
+ * Lock constructor.
+ *
+ * @param {Object} options
+ *   @property {String} redis Redis connection string.
+ *   @property {String=} namespace - An optional namespace under which to prefix all Redis keys and
+ *     channels used by this lock.
+ */
+function Lock(options) {
+  options = options || {};
+  options.namespace = options.namespace || 'lock';
+
+  this._namespace = options.namespace;
+
+  // Create Redis connection for issuing normal commands
+  this._redisConnection = redis.createClient(options.redis);
+
+  // Handler to run LUA scripts. Uses caching if possible
+  this._scripty = new Scripty(this._redisConnection);
+
+  // Redis connection with subscribers is not allowed to issue commands
+  // so we need an extra connection to handle subscription messages
+  this._redisSubscriber = redis.createClient(options.redis);
+
+  // Create event handler to register waiting locks
+  this._subscribers = new EventEmitter();
+  this._subscribers.setMaxListeners(Infinity);
+
+  // Whenever a lock is released it is published to the namespaced '-release' channel
+  // using the lock key as the message.
+  this._redisSubscriber.subscribe(`${this._namespace}-release`);
+  this._redisSubscriber.on('message', (channel, message) => {
+    if (channel !== `${this._namespace}-release` || !this._subscribers.listenerCount(message)) {
+      // just ignore, nothing to do here
+      return;
+    }
+
+    // Notify all waiting instances about the released lock
+    this._subscribers.emit(message);
+  });
+}
+
+Object.assign(Lock.prototype, {
+  /**
+   * Acquire a lock for a specific ID value. Callback returns the following value:
+   *
+   *   {
+   *     id: Message ID
+   *     success: either true (lock was acquired) of false (lock was not aquired)
+   *     state: state value for current lock
+   *     index: modification index for current lock
+   *     ttl: expiration time for the lock
+   *   }
+   *
+   * Lock index is a shared incrementing number (signed 64bit) that should ensure rogue
+   * lock holders would not be able to mess with newer locks for the same resource.
+   *
+   * @param {String} id Identifies the lock
+   * @param {String} state Informational text about the current state ('create', 'update', 'delete')
+   * @param {Number} ttl Automatically release lock after TTL (ms). Must be positive integer
+   * @param {Function} done Callback
+   */
+  acquireLock: function(id, state, ttl, done) {
+    var acquireScript = `
+        local ttl=tonumber(ARGV[2]);
+        if redis.call("EXISTS", KEYS[1]) == 1 then
+          local data = {
+            ["state"]=redis.call("HGET", KEYS[1], "state"),
+            ["index"]=tonumber(redis.call("HGET", KEYS[1], "index"))
+          };
+          return {0, data.state, data.index, redis.call("PTTL", KEYS[1])};
```

Comment:
It looks like this returns the index of the existing lock to the client, why would we want to do that? That would allow the client to release someone else's lock.